pkg/distributors/contracts/RewardsScheduler.sol

```diff
@@ -0,0 +1,134 @@
+        address indexed scheduler,
+        IERC20 indexed pool,
+        IERC20 indexed rewardsToken,
+        uint256 startTime,
+        uint256 amount
+    );
+    event RewardStarted(
+        bytes32 rewardId,
+        address indexed scheduler,
+        IERC20 indexed pool,
+        IERC20 indexed rewardsToken,
+        uint256 startTime,
+        uint256 amount
+    );
+
+    mapping(bytes32 => ScheduledReward) private _rewards;
+
+    function getScheduledRewardInfo(bytes32 rewardId) external view returns (ScheduledReward memory reward) {
+        return _rewards[rewardId];
+    }
+
+    function startRewards(bytes32[] calldata rewardIds) external {
+        for (uint256 r; r < rewardIds.length; r++) {
+            bytes32 rewardId = rewardIds[r];
+            ScheduledReward memory scheduledReward = _rewards[rewardId];
+
+            require(scheduledReward.status == RewardStatus.PENDING, "Reward cannot be started");
+            require(scheduledReward.startTime <= block.timestamp, "Reward start time is in the future");
+
+            scheduledReward.rewardsToken.approve(address(_multirewards), scheduledReward.amount);
+            _multirewards.notifyRewardAmount(
+                scheduledReward.pool,
+                scheduledReward.rewardsToken,
+                scheduledReward.amount,
+                scheduledReward.rewarder
+            );
+            emit RewardStarted(
+                rewardId,
+                msg.sender,
+                scheduledReward.pool,
+                scheduledReward.rewardsToken,
+                scheduledReward.startTime,
+                scheduledReward.amount
+            );
+
+            _rewards[rewardId].status = RewardStatus.STARTED;
+        }
+    }
+
+    function getRewardId(
+        IERC20 pool,
+        IERC20 rewardsToken,
+        address rewarder,
+        uint256 startTime
+    ) public pure returns (bytes32) {
+        return keccak256(abi.encodePacked(pool, rewardsToken, rewarder, startTime));
+    }
+
+    function scheduleReward(
+        IERC20 pool,
+        IERC20 rewardsToken,
+        uint256 amount,
+        uint256 startTime
+    ) public returns (bytes32 rewardId) {
+        rewardId = getRewardId(pool, rewardsToken, msg.sender, startTime);
+        require(startTime > block.timestamp, "Reward can only be scheduled for the future");
+        require(
+            _multirewards.isAllowlistedRewarder(pool, rewardsToken, msg.sender),
+            "Only allowlisted rewarders can schedule reward"
+        );
+
+        require(_rewards[rewardId].status == RewardStatus.UNINITIALIZED, "Reward has already been scheduled");
+
+        rewardsToken.safeTransferFrom(msg.sender, address(this), amount);
+
+        _rewards[rewardId] = ScheduledReward({
+            pool: pool,
+            rewardsToken: rewardsToken,
+            rewarder: msg.sender,
+            amount: amount,
+            startTime: startTime,
+            status: RewardStatus.PENDING
+        });
```

Comment:
As in the other comment, I'd initialize the mapping before calling `safeTransferFrom` to avoid reentrancy issues.